addons/libs/dialog.lua

```diff
@@ -0,0 +1,165 @@
+-- This library was written to help find the ID of a known
+-- action message corresponding to an entry in the dialog tables.
+-- While the IDs can be collected in-game, they occasionally
+-- change and would otherwise need to be manually updated.
+-- It can also be used to find and decode an entry given the ID.
+
+-- Common parameters:
+-- dat: The entire content of the zone dialog DAT file
+-- i.e. local dat = io.open('path/to/dialog/DAT', 'rb'):read('*a')
+-- entry: The string you are looking for. If you do not know the
+-- entire string, use dev_find_substring.
+
+local xor = require('bit').bxor
+local floor = require('math').floor
+local string = require('string')
+local find = string.find
+local sub = string.sub
+local gsub = string.gsub
+local format = string.format
+local char = string.char
+local byte = string.byte
+require('pack')
+local unpack = string.unpack
+local pack = string.pack
+
+local dialog = {}
+
+local function decode(int)
+    return xor(int, 0x80808080)
+end
+local encode = decode
+
+local function binary_search(pos, dat, n)
+    local l, r, m = 1, n
+    while l < r do
+        m = floor((l + r) / 2)
+        if decode(unpack('<I', dat, 1 + 4 * m)) < pos then
+        -- offset given by mth ID < offset to string
+            l = m + 1
+        else
+            r = m
+        end
+    end
+    return l - 2 -- we want the index to the left of where "pos" would be placed
+end
+
+local function plain_text_gmatch(text, substring, n)
+    n = n or 1
+    return function()
+        local pos = find(text, substring, n, true)
+        if pos then n = pos + 1 end
+        return pos
+    end
+end
+
+-- This function may help if you are viewing the decoded contents
+-- of a dialog DAT within a hex editor.
+function dialog.dev_get_offset(dat, id)
+    return format('%x', decode(unpack('<I', dat, 5 + 4 * id)))
+end
+
+-- Returns the number of entries in the given dialog DAT file
+function dialog.entry_count(dat)
+    return decode(unpack('<I', dat, 5)) / 4
+end
+
+-- Returns an array-like table containing every ID which matched
+-- the given entry.
+-- An important note about this function: The tables contain an
+-- enormous number of duplicate entries. Additionally, some short
+-- entries can occur as substrings of longer entries. Results
+-- which fall into the second case should be filtered out.
+-- If you are dealing with the first case, it may help to look at
+-- any other messages which were received at the same time.
+function dialog.get_ids_matching_entry(dat, entry)
+    local last_offset = decode(unpack('<I', dat, 5))
+    local res = {}
+    local n = 0
+    local start = 5
+    for i in plain_text_gmatch(dat, entry, last_offset) do
```

Comment:
Without having tested it myself, are the strings in the DAT not encoded? Or why would a plain text search work?

Reply:
They are encoded. The "plain text" was meant to refer to the "plain" parameter for `string.find`. It searches for the given text, ignoring patterns.
I might go back through later next weekend to see if escaping the string and using string.gmatch is faster.

Reply:
No, I think this is the better way, as discussed on Discord. I just meant that if I search for `Hello`, how would it ever match anything, if the file is encoded? It would not appear as `Hello` in the file, which is `\x48\x65\x6C\x6C\x6F`, but as `\xC8\xE5\xEC\xEC\xEF`. So the match would not return any hits, or am I missing something?

Reply:
I updated the description for the `entry` parameter. It's meant to be the encoded string, so you'd search for `dialog.encode('Hello')`.
That's a little inconvenient on the surface, but in practice I don't think anyone will be typing the entries in. They contain so many special bytes that they're usually unreadable even after decoding. My thought was that they would get the entry either with the ID using `get_entry` or via `dev.find_substring` and then use `dialog.serialize` to get a version of the string they could work with.
I'll probably switch `dev.find_substring` around so that it uses the decoded string, since I can see it being used in that way.